### PR TITLE
Refactors kubernetes resource templating to task interface

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Towards https://github.com/giantswarm/architect/issues/73

### CODE REVIEW

File: template/error.go [hunk 1]
```diff
@@ -45,3 +45,10 @@ var incorrectTemplatingError = errgo.New("incorrect templating")
 func IsIncorrectTemplating(err error) bool {
 	return errgo.Cause(err) == incorrectTemplatingError
 }
+
+var emptyKubernetesResourcesDirectoryPath = errgo.New("empty kubernetes resources directory path")
```
Comment (xh3b4sd) [inline template/error.go]: `emptyDirectoryPathError` would work for me. You could then mask the error with additional information ala `microerror.MaskAnyf(emptyDirectoryError, "kubernetes resource '%s'", thePath)`.

File: template/kubernetes.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+func (t TemplateKubernetesResourcesTask) Run() error {
+	if t.kubernetesResourcesDirectoryPath == "" {
+		return microerror.MaskAny(emptyKubernetesResourcesDirectoryPath)
+	}
+
+	if err := utils.CopyDir(
+		t.fs,
+		t.kubernetesResourcesDirectoryPath,
+		t.templatedResourcesDirectoryPath,
+	); err != nil {
```
Comment (xh3b4sd) [inline template/kubernetes.go]: Every time function calls and conditionals are mixed a pomsky puppy dies. Please don't kill the puppies.

File: template/kubernetes.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+// templateKubernetesResources takes a filesystem,
+// a path to a directory holding kubernetes resources,
+// and an installation configuration.
+// It templates the given resources, with data from the configuration,
+// writing changes to the files.
+func templateKubernetesResources(fs afero.Fs, resourcesPath string, config TemplateConfiguration) error {
+	walkFunc := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			microerror.MaskAny(err)
+		}
```
Comment (xh3b4sd) [inline template/kubernetes.go]: I did this myself recently and wondered what the use case for this propagated error is. This feels super weird.

File: template/kubernetes.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+		var buf bytes.Buffer
+		if err := t.Execute(&buf, config); err != nil {
+			microerror.MaskAny(err)
+		}
+
+		templatedContents := buf.String()
+
+		// This add backwards compatability for `%%DOCKER_TAG%%`. Deprecated.
```
Comment (xh3b4sd) [inline template/kubernetes.go]: He, she it, das "S" muss mit. :D

File: template/kubernetes.go (hunk 2 above)
Comment (JosephSalisbury) [inline template/kubernetes.go]: I know we disagree with
```
if err := func(); err != nil {
```

gonna go with this :D

File: template/kubernetes.go (hunk 3 above)
Comment (JosephSalisbury) [inline template/kubernetes.go]: I've never even seen this error actually be anything except for `nil`. Feels wrong not to handle it, but I don't think it's a big deal.

File: template/kubernetes.go (hunk 4 above)
Comment (JosephSalisbury) [inline template/kubernetes.go]: ja ja genau

Comment (JosephSalisbury) [inline template/kubernetes.go]: fixeden

File: template/error.go (hunk 1 above)
Comment (JosephSalisbury) [inline template/error.go]: no strong feelings. this error will go pretty soon when we remove the push deployments support

**Pull request merged**